shaping/render_test.go

```diff
@@ -0,0 +1,182 @@
+		for _, run := range line {
+			if w := run.GlyphBounds.LineThickness().Round(); w > width {
+				width = w
+			}
+			height += -run.Advance.Round()
+		}
+	} else {
+		baseline = firstRun.GlyphBounds.Ascent.Round()
+		for _, run := range line {
+			if h := run.GlyphBounds.LineThickness().Round(); h > height {
+				height = h
+			}
+			width += run.Advance.Round()
+		}
+	}
+	return
+}
+
+func drawTextLine(runs []Output, file string) error {
+	width, height, baseline := imageDims(runs)
+	img := image.NewRGBA(image.Rect(0, 0, width, height))
+	// white background
+	draw.Draw(img, img.Rect, image.NewUniform(color.White), image.Point{}, draw.Src)
+
+	if runs[0].Direction.IsVertical() {
+		// draw the baseline
+		drawVLine(img, image.Pt(baseline, 0), height, black)
+
+		dot := image.Pt(baseline, 0)
+		for _, run := range runs {
+			dot = drawVRun(run, img, dot)
+		}
+	} else {
+		// draw the baseline
+		drawHLine(img, image.Pt(0, baseline), width, black)
+
+		dot := image.Pt(0, baseline)
+		for _, run := range runs {
+			dot = drawHRun(run, img, dot)
+		}
+	}
+
+	f, err := os.Create(file)
+	if err != nil {
+		return err
+	}
+	if err = png.Encode(f, img); err != nil {
+		return err
+	}
+	err = f.Close()
+	return err
+}
+
+// assume horizontal direction
+func drawHRun(out Output, img *image.RGBA, dot image.Point) image.Point {
+	sizeFactor := float32(out.Size.Round()) / float32(out.Face.Upem())
+	for _, g := range out.Glyphs {
+		minX := dot.X + g.XOffset.Round() + g.XBearing.Round()
+		maxX := minX + g.Width.Round()
+		minY := dot.Y + g.YOffset.Round() - g.YBearing.Round()
+		maxY := minY - g.Height.Round()
+
+		drawRect(img, image.Pt(minX, minY), image.Pt(maxX, maxY), green)
+
+		// draw the dot
+		drawPoint(img, dot, black)
+
+		// draw a sketch of the glyphs
+		glyphData := out.Face.GlyphData(g.GlyphID).(api.GlyphOutline)
+		drawGlyph(img, dot.Add(image.Pt(g.XOffset.Ceil(), g.YOffset.Ceil())), glyphData, sizeFactor, black)
+
+		dot.X += g.XAdvance.Round()
+		// draw the advance
+		drawVLine(img, image.Pt(dot.X, 0), img.Bounds().Dy(), red)
+	}
+
+	return dot
+}
+
+// assume vertical direction
+func drawVRun(out Output, img *image.RGBA, dot image.Point) image.Point {
+	sizeFactor := float32(out.Size.Round()) / float32(out.Face.Upem())
+	for _, g := range out.Glyphs {
+		dot.Y += -g.YAdvance.Round()
+
+		minX := dot.X + g.XOffset.Round() + g.XBearing.Round()
+		maxX := minX + g.Width.Round()
+
+		minY := dot.Y + g.YOffset.Round() - g.YBearing.Round()
+		maxY := minY - g.Height.Round()
+
+		drawRect(img, image.Pt(minX, minY), image.Pt(maxX, maxY), green)
+
+		// draw the dot
+		drawPoint(img, dot, black)
+
+		// draw a sketch of the glyphs
+		glyphData := out.Face.GlyphData(g.GlyphID).(api.GlyphOutline)
+		if out.Direction.IsSideways() {
+			glyphData.Sideways(float32(-g.YAdvance.Round()) / sizeFactor)
+		}
+		drawGlyph(img, dot.Add(image.Pt(g.XOffset.Ceil(), g.YOffset.Ceil())), glyphData, sizeFactor, black)
```

Comment:
Question: why isn't this `dot.Add(image.Pt(g.XOffset.Ceil(), -g.YOffset.Ceil()))`?

Reply:
Oh, I _think_ you are correct here ! I'll double check tomorrow